sdk/core/azure-core/src/main/java/com/azure/core/util/ProgressReceiver.java

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.util;
+
+/**
+ * A {@link ProgressReceiver} is an interface that can be used to report progress on network transfers.
+ * The {@link #reportProgress(long)} method will be called periodically with the total number of bytes
+ * transferred.
+ */
+@FunctionalInterface
+public interface ProgressReceiver {
+    /**
+     * The callback function invoked as progress is reported.
+     *
+     * <p>
+     * The callback can be called concurrently from multiple threads if reporting spans across multiple
+     * requests.
+     * </p>
+     *
+     * @param bytesTransferred The total number of bytes transferred during this transaction.
+     */
+    void reportProgress(long bytesTransferred);
```

Comment:
Yeah. we can definietly change name of the parameter to something more generic.

When it comes to symbols. I had `ProgressListener.onProgress` before but in offline discussion with @JonathanGiles we figured that diverging too much from existing pattern is going to cause user pain and noise on existing types like options bags that take progressreceivers today if they have implementations based on existing types.
1. We want to match signature of existing `ProgressReceiver`s like this one https://github.com/kasobol-msft/azure-sdk-for-java/blob/36b58fde61222042d7915b88a92313633aacb819/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ProgressReceiver.java#L27 and make them extend new core type https://github.com/kasobol-msft/azure-sdk-for-java/blob/36b58fde61222042d7915b88a92313633aacb819/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/ProgressReceiver.java#L20 . The reason for that is good deprecation story when they have implementations based on existing interfaces and we stop developing APIs based on old interfaces and deprecate them eventually.
2. As for the type name we could do `ProgressReporter` and rename the other `ProgressReporter` (the one that is called by our impl to publish) to `ProgressCollector`/`ProgressPublisher` or so.


Or perhaps if we want to revisit naming we could make something like
```
core.ProgressListener {
 void onProgress(progress);
}

storage.ProgressReceiver {
 void reportProgress(progress);
 default onProgress(progress) {
   //i.e. delegate to old abstraction by default.
   reportProgress(progress);
 }
}

``

Please share your thoughts.

Reply:
Yeah, I think `ProgressReporter` seems like a more accurate name for what this interface does (reporting progress). For the other type, I like `ProgressPublisher` but `Publisher` is overloaded with reactive stream publishers. Other names to consider:
- ProgressTracker
- ProgressCollector (mentioned above)

Reply:
@JonathanGiles Please chime in on naming here.

Reply:
I think the choices discussed above can boil down to either ProgressReporter/reportProgress, ProgressListener/onProgress, or ProgressReceiver/reportProgress. From the standpoint of a developer who is providing this interface implementation, they are expecting to listen to (or receive) progress, rather than report progress. In this interpretation, I think we should rule out `ProgressReporter`. That leaves us with `ProgressListener` or `ProgressReceiver`. I don't have a strong preference here, although I would lean towards a `Listener` being more commonly understood by developers as a general term. In terms of `onProgress` vs `reportProgress`, I again think from a consumers point of view, they are not `reporting progress`, and we should probably name the API from the consuming users standpoint, so `onProgress` or `receiveProgress` or `consumeProgress` or `consume` or anything along these lines should work fine.

Reply:
I sympathize with @JonathanGiles 's pov on naming. However.

Are we fine with this (and it's copies) 
https://github.com/Azure/azure-sdk-for-java/blob/0a1b960a452f853c7cde7c309ab0b41eddfa7b52/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/ProgressReceiver.java#L17-L25

becoming
```java
package com.azure.storage.common;

// Eventually get rid of this type.
public interface ProgressReceiver extends ProgressListener {

    @Deprecated
    void reportProgress(long bytesTransferred);
    
    default void onProgress(long progress) {
       reportProgress(progress);
    }
}
```

??

Reply:
Yes, I'm fine with that (although it'll be `handleProgress` rather than `onProgress`, unless I get overruled).

Reply:
If we're going with `ProgressListener.handleProgress` then I'll keep `ProgressReporter.reportProgress` as is.

Reply:
Sample deprecation https://github.com/kasobol-msft/azure-sdk-for-java/commit/460f7b24ccb5f5ddba183568464a76b96893bbfb .